### PR TITLE
Update RPC framework

### PR DESCRIPTION
**Pull request by lvyelin**
## Issue

Update RPC framework

## Changes:

1. Remove eth rpc code and unnecessary changes from
7a0f18f. Import the package instead.
2. Move HmyAPIBackend into core package so it has more access to core.
3. Add API interface for services.
4. Start RPC for all roles, for both HTTP (baseport+10) and WS (baseport+20).
3. Keep implemented APIs but move to internal/hmyapi package.

## Test
$ curl -H "Content-Type: application/json" -d '{"method":"hmy_getBlockByNumber","params":["latest",true],"id":1}' http://127.0.0.1:9010/
{"jsonrpc":"2.0","id":1,"result":{"difficulty":"0x0","extraData":"0x","gasLimit":"0x35b7ffdfe","gasUsed":"0x17f00","hash":"0x4ac90cba6c09496855a46139dfe36df1c9e54e90e420eb5944d69953a4bf3aa7","logsBloom":"0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000","miner":"0x0000000000000000000000000000000000000000","mixHash":"0x0000000000000000000000000000000000000000000000000000000000000000","nonce":"0x0000000000000000","number":"0x177","parentHash":"0x18a8a5e6f10d3e368e16c0de6ad287247c45eb3e0cb405a78f3352c66000b3d2","receiptsRoot":"0x9a4f1f141492f2870a8bf8fa6e8646fcb6007bf66a71c8e7687d4dda7ac5c057","size":"0x3ab","stateRoot":"0x86bf5ac1f6d8385640901968ebc3893f2500cbb9edf454b4fbb9537c96772d95","timestamp":1557723465,"transactionsRoot":"0x9a24d322060f456d1f146dcb064dd3c2b0715305b30c9560e838911023c704b7","uncles":[]}}

### CODE REVIEW

**Comment by LeoHChen** [comment]
The PR failed travis-ci, you need to run golint to fix the format issue.

**Comment by lvyelin** [comment]
@LeoHChen @lzl124631x 
Update the patch and keep APIs previously implemented.
Please review.

File: internal/hmyapi/backend.go [hunk 1]
```diff
@@ -1,22 +1,21 @@
...
-const namespace = "hmy"
-
 // GetAPIs returns all the APIs.
-func GetAPIs(b *rpc.HmyAPIBackend) []rpc.API {
+func GetAPIs(b *core.HmyAPIBackend) []rpc.API {
 	nonceLock := new(AddrLocker)
 	return []rpc.API{
 		{
-			Namespace: namespace,
+			Namespace: "hmy",
```
Comment (LeoHChen) [inline internal/hmyapi/backend.go]: why not use const?

File: node/rpc.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	httpListener net.Listener
+	httpHandler  *rpc.Server
+
+	wsListener net.Listener
+	wsHandler  *rpc.Server
+
+	httpEndpoint = ""
+	wsEndpoint   = ""
+
+	httpModules      = []string{"hmy"}
```
Comment (LeoHChen) [inline node/rpc.go]: const?

File: node/rpc.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	apiBackend = core.NewBackend(node.blockchain, node.TxPool)
+
+	apis := hmyapi.GetAPIs(apiBackend)
+	for _, service := range node.serviceManager.GetServices() {
+		apis = append(apis, service.APIs()...)
+	}
+
+	port, _ := strconv.Atoi(nodePort)
+
+	httpEndpoint = fmt.Sprintf("localhost:%v", port+rpcHTTPPortOffset)
```
Comment (LeoHChen) [inline node/rpc.go]: use net.JoinHostPort, and why localhost?

File: node/rpc.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	if err := node.startHTTP(httpEndpoint, apis, httpModules, nil, httpVirtualHosts, httpTimeouts); err != nil {
+		return err
+	}
+
+	wsEndpoint = fmt.Sprintf("localhost:%v", port+rpcWSPortOffset)
```
Comment (LeoHChen) [inline node/rpc.go]: same as above, why localhost?

File: node/rpc.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+
+	"github.com/ethereum/go-ethereum/log"
```
Comment (LeoHChen) [inline node/rpc.go]: use our internal log?

File: api/service/restclientsupport/service.go [hunk 6]
```diff
@@ -259,3 +260,8 @@ func (s *Service) NotifyService(params map[string]interface{}) {
 func (s *Service) SetMessageChan(messageChan chan *msg_pb.Message) {
 	s.messageChan = messageChan
 }
+
+// APIs for the services.
+func (s *Service) APIs() []rpc.API {
```
Comment (lzl124631x) [inline api/service/restclientsupport/service.go]: All the APIs are nil. I don't think it's necessary to add this interface.

File: core/api_backend.go [hunk 7]
```diff
@@ -1,30 +1,30 @@
-package rpc
+package core
```
Comment (lzl124631x) [inline core/api_backend.go]: I feel `api_backend` is only serving rpc, so putting it into rpc folder makes the code more cohesive.

Comment (lzl124631x) [inline core/api_backend.go]: I feel `api_backend` is only serving/relevant to rpc (not shared/used by anything else), so putting it into rpc folder makes the code more cohesive.

File: api/service/restclientsupport/service.go (hunk 6 above)
Comment (LeoHChen) [reply api/service/restclientsupport/service.go]: my understanding is all services can provide RPC service, via this interface.  sounds like a good framework.

File: internal/hmyapi/backend.go (hunk 1 above)
Comment (lvyelin) [reply internal/hmyapi/backend.go]: may add some other API with different namespace. so didn't use const.

File: node/rpc.go (hunk 2 above)
Comment (lvyelin) [reply node/rpc.go]: we may able to modify enabled modules at run time and restart RPC to provide more modules. so not using const.

File: api/service/restclientsupport/service.go (hunk 6 above)
Comment (lvyelin) [reply api/service/restclientsupport/service.go]: Yes, it is just a framework, so the services can provide API.

File: core/api_backend.go (hunk 7 above)
Comment (lvyelin) [reply core/api_backend.go]: Moving to core package can access more in the package.
It is the functions that core provided to the RPC, so it should be put into the provider (core package).

**Comment by lzl124631x** [comment]
Thanks for the change. My bad that I didn't realize that we don't at all need to copy & paste eth-rpc code. Good catch.

It looks good overall.

**Comment by lvyelin** [comment]
@LeoHChen @lzl124631x Do you agree to use +10 and +20 ports?

File: node/rpc.go [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/internal/hmyapi"
+)
+
+const (
+	rpcHTTPPortOffset = 10
+	rpcWSPortOffset   = 20
```
Comment (lzl124631x) [inline node/rpc.go]: Move all port offsets to a single file so that we won't step on each other.

**Comment by lzl124631x** [comment]
> @LeoHChen @lzl124631x Do you agree to use +10 and +20 ports?

Sure. Nit: Move all port offsets to a single file so that we won't step on each other and we can know the ports at a glance without needing to jump between files.

File: node/rpc.go (hunk 5 above)
Comment (lvyelin) [reply node/rpc.go]: I see our code also use eth log. which log do you prefer?

File: api/service/restclientsupport/service.go (hunk 6 above)
Comment (lzl124631x) [reply api/service/restclientsupport/service.go]: I can see your point. But I don't think it's needed in this PR -- kinda premature to add it.
* If we found we never add APIs in this way these interfaces are dead code. 
* If we need to add APIs to some service, why not add it by then.

File: internal/hmyapi/backend.go (hunk 1 above)
Comment (lzl124631x) [reply internal/hmyapi/backend.go]: But for this hmy namespace I still think we should make it a const.

File: core/api_backend.go (hunk 7 above)
Comment (lzl124631x) [reply core/api_backend.go]: Ah. You're right I confused this with the internal public.go. This one matches better with `eth/api_backend.go`

File: node/rpc.go (hunk 8 above)
Comment (LeoHChen) [reply node/rpc.go]: there is a ticket for this task.  https://github.com/harmony-one/harmony/issues/612, no need to block this PR.

File: node/rpc.go (hunk 5 above)
Comment (LeoHChen) [reply node/rpc.go]: try https://github.com/harmony-one/harmony/commit/83bd41fb930dd3822f2ceb9d3b22bbde7dbc375f

**Pull request merged**